File: cc/001-010/004.cc
// Project Euler Problem 004
//
// Find the largest palindrome made from the product of two 3-digit numbers.

#include <cstdio>

int palindrome(int n) {
  int t = n;
  int sum = 0;
  while (t) {
    sum = sum * 10 + (t % 10);
    t = t / 10;
  }
  return (sum == n);
}

int main(char argv[]) {
  int max = 0;
  for (int i = 999; i > 99; i--) {
    if (999 * i <= max) {
      break;
    }
    for (int j = 999; j >= i; j--) {
      int product = i * j;
      if (product <= max) {
        break;
      }
      if (palindrome(product)) {
        max = product;
      }
    }
  }
  printf("%d\n", max);
}

File: go/problems/004.go
// Project Euler Problem 004
//
// Find the largest palindrome made from the product of two 3-digit numbers.

package problems

import "fmt"

func isPalindrome(n int) bool {
	s := fmt.Sprint(n)
	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
		if s[i] != s[j] {
			return false
		}
	}
	return true
}

func Problem004() string {
	max := 0
	for i := 101; i <= 999; i++ {
		for j := i; j <= 999; j++ {
			s := i * j
			if s > max && isPalindrome(s) {
				max = s
			}
		}
	}
	return fmt.Sprint(max)
}

func init() {
	Functions[4] = Problem004
}
